src/utils/openmc/universe.py

```diff
@@ -1058,7 +1073,70 @@ def set_universes(self, universes):
         # The sub-lists are ordered from outermost ring to innermost ring.
         # The Universes within each sub-list are ordered from the "top" in a
         # clockwise fashion.
-        self.set_num_rings(universes.shape[0])
+
+        # Check to see if the given universes look like a 2D or a 3D array.
+        if isinstance(self._universes[0][0], Universe):
+            n_dims = 2
+
+        elif isinstance(self._universes[0][0][0], Universe):
+            n_dims = 3
+
+        else:
+            msg = 'HexLattice ID={0:d} does not appear to be either 2D or ' \
+                  '3D.  Make sure set_universes was given a two-deep or ' \
+                  'three-deep iterable of universes.'.format(self._id)
+            raise RuntimeError(msg)
+
+        # Set the number of axial positions.
+        if n_dims == 3:
+            self.set_num_axial(self._universes.shape[0])
+
+        else:
+            self._num_axial = None
+
+        # Set the number of rings and make sure this number is consistent for
+        # all axial positions.
+        if n_dims == 3:
+            self.set_num_rings(len(self._universes[0]))
+            for rings in self._universes:
+                if len(rings) != self._num_rings:
+                    msg = 'HexLattice ID={0:d} has an inconsistent number of ' \
+                          'rings per axial positon'.format(self._id)
+                    raise ValueError(msg)
+
+        else:
+            self.set_num_rings(self._universes.shape[0])
+
+        # Make sure there are the correct number of elements in each ring.
+        z = 0
+        while True:
```

Comment:
I don't tend to like `` while True`` loops since they are prone to becoming infinite while loops in future extensions of the code. And in particular, it doesn't look like the second iteration of this ``while`` loop will execute any differently than the first iteration, and so on, which means it is a perfect trap for infinite loops. Looking at the code block, I think you can break this up in such a way that it is just a series of error checking cases. Can you do this, or do you have a good reason to do it inside of a loop?